File: ruby/lib/epostak/resources/box.rb
# frozen_string_literal: true

require "erb"

module EPostak
  module Resources
    # ePošťák Box durable execution layer for staged, scheduled, and retryable
    # Peppol dispatch.
    class Box
      def initialize(http)
        @http = http
      end

      def list(status: nil, direction: nil, limit: nil, offset: nil)
        @http.request(
          :get,
          "/box/items",
          query: { status: status, direction: direction, limit: limit, offset: offset },
        )
      end

      def create(payload_xml:, scheduled_for: nil, external_id: nil, metadata: nil)
        @http.request(
          :post,
          "/box/items",
          body: {
            payloadXml: payload_xml,
            scheduledFor: scheduled_for,
            externalId: external_id,
            metadata: metadata,
          }.compact,
        )
      end

      def get(item_id)
        @http.request(:get, "/box/items/#{encode(item_id)}")
      end

      def schedule(item_id, scheduled_for:)
        @http.request(
          :post,
          "/box/items/#{encode(item_id)}/schedule",
          body: { scheduledFor: scheduled_for },
        )
      end

      def send_now(item_id)
        @http.request(:post, "/box/items/#{encode(item_id)}/send-now")
      end

      def retry(item_id)
        @http.request(:post, "/box/items/#{encode(item_id)}/retry")
      end

      def cancel(item_id)
        @http.request(:post, "/box/items/#{encode(item_id)}/cancel")
      end

      private

      def encode(value)
        ERB::Util.url_encode(value.to_s)
      end
    end
  end
end

File: python/src/epostak/resources/box.py
"""ePostak Box resource -- durable, scheduled, and retryable dispatch layer."""

from __future__ import annotations

from typing import Any, Dict, Optional, TYPE_CHECKING, Union
from urllib.parse import quote

from epostak.resources.documents import _BaseResource, _build_query

if TYPE_CHECKING:
    from epostak.types import (
        BoxItem,
        BoxItemDetail,
        BoxCreateRequest,
        BoxListResponse,
        BoxScheduleRequest,
    )


class BoxResource(_BaseResource):
    """ePošťák Box durable execution layer for Peppol dispatch."""

    def list(
        self,
        *,
        status: Optional[str] = None,
        direction: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> "BoxListResponse":
        params = _build_query({
            "status": status,
            "direction": direction,
            "limit": limit,
            "offset": offset,
        })
        return self._request("GET", "/box/items", params=params)

    def create(self, body: "BoxCreateRequest") -> Dict[str, Any]:
        return self._request("POST", "/box/items", json=body)

    def get(self, item_id: str) -> "BoxItemDetail":
        return self._request("GET", f"/box/items/{quote(item_id, safe='')}")

    def schedule(
        self,
        item_id: str,
        scheduled_for: Union[str, "BoxScheduleRequest"],
    ) -> "BoxItem":
        body: Dict[str, Any]
        if isinstance(scheduled_for, dict):
            body = dict(scheduled_for)
        else:
            body = {"scheduledFor": scheduled_for}
        return self._request(
            "POST",
            f"/box/items/{quote(item_id, safe='')}/schedule",
            json=body,
        )

    def send_now(self, item_id: str) -> Dict[str, Any]:
        return self._request("POST", f"/box/items/{quote(item_id, safe='')}/send-now")

    def retry(self, item_id: str) -> "BoxItem":
        return self._request("POST", f"/box/items/{quote(item_id, safe='')}/retry")

    def cancel(self, item_id: str) -> "BoxItem":
        return self._request("POST", f"/box/items/{quote(item_id, safe='')}/cancel")
